UtilityModules/src/StopSelection_module.cc

```diff
@@ -0,0 +1,290 @@
+                               std::unique_ptr<StepPointMCCollection>& out_steps,
+                               art::ProductID const& pid, art::EDProductGetter const* getter) {
+    // for every input step point, add it to the output if relevant, with remapping
+    for(auto& step : steps) {
+      bool found = false;
+      const auto id = step.simParticle()->id();
+      for(auto& sim : *sims) {
+        if(sim.first == id) {
+          found = true;
+          break;
+        }
+      }
+      if(found) { // save the step
+        auto out_step(step);
+        out_step.simParticle() = art::Ptr<SimParticle>(pid, step.simParticle().key(), getter);
+        out_steps->push_back(out_step);
+      }
+    }
+  }
+
+  //--------------------------------------------------------------------------------------------------------------
+  bool StopSelection::filter(art::Event& event) {
+    auto out_sims{std::make_unique<SimParticleCollection>()};
+    auto out_steps{std::make_unique<StepPointMCCollection>()};
+    const PhysicsParams& gc = *GlobalConstantsHandle<PhysicsParams>();
+    const bool make_weight = !decayOffPdgs_.empty();
+    double weight = 1.;
+
+    // for mapping new sim art Ptrs
+    auto out_pid = event.getProductID<SimParticleCollection>("");
+    auto out_getter = event.productGetter(out_pid);
+
+    // get the sim collection and the corresponding stops
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_);
+    const auto steps = event.getValidHandle<StepPointMCCollection>(stepsToken_);
+    const auto stops = simParticleList(simh, pdgId_, processCode_); // stopped particles
+    const auto selected = pruneStops(stops); // stopped particles satisfying a given selection
+
+    if(diagLevel_ > 1) {
+      printf("[StopSelection::%s] From %zu stops selected %zu candidates\n", __func__, stops.size(), selected.size());
+    }
+
+    const size_t nstops = selected.size();
+    const int offset = (acceptReject_) ? 0 : randomFlat_.fireInt(nstops); // if picking 1 stop, make it random, otherwise sample all
+    for(size_t istop = 0; istop < nstops; ++istop) {
+      const size_t index = (istop + offset) % nstops;
+      art::Ptr<SimParticle> sim = selected[index];
+
+      // get the selected stop's time weight if requested
+      if(make_weight) {
+        const double tau = SimParticleGetTau::calculate(sim, decayOffPdgs_, gc);
+        weight = std::exp(-tau);
+      }
+      total_.add(weight);
+
+      // check if doing accept/reject
+      if(acceptReject_) {
+        if(weight < maxWeight_ * randomFlat_.fire()) continue; // if it fails, continue to the next stop
+        weight = 1.; // reset as no event weight is used if using accept/reject
+      }
+
+      // save the entire lineage, updating the pointer mapping
+      addSim(sim, out_sims, out_pid, out_getter);
+
+      selected_.add(weight);
+      if(!acceptReject_) break; // no need to check each stop if not doing accept/reject
```

Comment:
Will the weight sum be correct if you exit the nstop loop early if not running accept/reject?

Reply:
I just pushed a commit with this additional sum of weights output (one for the stops considered in the sampling, one for the total of the eligible stop list, independent of stops considered). I think this is the last code issue, so I can next move onto the Production PR to update the simulation workflow I believe